### PR TITLE
Update usages of IWantToRunWhenBusStartsAndStops in snippets and upgrade guides

### PR DESCRIPTION
**Pull request by colin-higgins**
Connects to https://github.com/Particular/NServiceBus/issues/3342

`IWantToRunWhenBusStartsAndStops` moved from core to hosts as `IWantToRunWhenEndpointStartsAndStops`

FeatureStartupTask
 - [x] Documentation
   - [x] \nservicebus\upgrades\4to5.md
   - [x] \nservicebus\upgrades\5to6.md
 - [x] nservicebus\lifecycle\endpointstartandstop.md
   - [x] Clarify interface has moved to host for v6
   - [x] Link to self host / link to feature startup task

PING @kbaley @boblangley

### CODE REVIEW

**Comment by colin-higgins** [comment]
I think we need to do a lot of the behavior documentation in snippets between v3/4/5/6

I don't see any other good way to document the behavior differences without taking up a massive amount of real estate.

**Comment by boblangley** [comment]
I updated our description with a more detailed description of the changes needed.

**Comment by boblangley** [comment]
For reference: Documentation link for FeatureStartupTask http://docs.particular.net/nservicebus/pipeline/features#feature-startup-tasks

**Comment by boblangley** [comment]
I pushed the snippet split and updated the v6 lifecycle snippet to show self-hosted options

File: Snippets/Snippets_6/Lifecycle/EndpointStartAndStop.cs [hunk 1]
```diff
@@ -2,19 +2,48 @@
...
             // perform startup logic
+
+            await endpoint.Stop();
+
+            // perform shutdown logic
         }
+    }
 
-        public async Task Stop(IMessageSession session)
+    class RunWhenEndpointStartsAndStopsFeature : Feature
```
Comment (kbaley) [inline Snippets/Snippets_6/Lifecycle/EndpointStartAndStop.cs]: We might not want to talk about Features as it's required only if you need DI and are using the NSB container and not your own.

Comment (boblangley) [inline Snippets/Snippets_6/Lifecycle/EndpointStartAndStop.cs]: I couldn't remember where we landed on including this, so I put it here anyhow. The only reason for including it is for users that are using our container, which we are discouraging.

**Comment by boblangley** [comment]
We should probably move the Host snippets out of the main snippets projects and to a sub folder of projects specific to the Hosts. This would allow the snippet "versions" to match the Host version (instead of core). If you look RabbitMQ and Azure Service Bus has done something similar.

File: nservicebus/hosting/critical-errors.md [hunk 2]
```diff
@@ -13,7 +13,7 @@ For many scenarios NServiceBus has built-in [error and exception management](/ns
  * An Exception occurs when NServiceBus is attempting to move a message to the Error Queue.
  * There are repeated failures in reading information from a required storage.
  * An exception occurs reading from the input queue.
- * A `IWantToRunWhenBusStartsAndStops.Start` throws an exception.
+ * A `IWantToRunWhenEndpointStartsAndStops.Start` throws an exception.
```
Comment (SimonCropp) [inline nservicebus/hosting/critical-errors.md]: this will make it invalid for older versions

File: nservicebus/upgrades/5to6.md [hunk 3]
```diff
@@ -660,4 +660,10 @@ NOTE: This ID needs to be stable and should never be hardcoded, e.g. it can be r
...
 
-snippet: 5to6-ExplicitProperties
\ No newline at end of file
+snippet: 5to6-ExplicitProperties
+
+## IWantToRunWhenBusStartsAndStops is obsolete
+
+The `IWantToRunWhenBusStartsAndStops` interface is no longer supported. When self-hosting, call any startup code after `Endpoint.Create` or any cleanup code after `Endpoint.Stop`. When using either NServiceBus.Host or NServiceBus.Host.AzureCloudService, use the host's `IWantToRunWhenEndpointStartsAndStops` interface instead.
+
+For tests that implement `IWantToRunWhenBusStartsAndStops`, use a `FeatureStartupTask` instead.
```
Comment (SimonCropp) [inline nservicebus/upgrades/5to6.md]: link?

File: nservicebus/upgrades/5to6.md [hunk 4]
```diff
@@ -660,4 +660,10 @@ NOTE: This ID needs to be stable and should never be hardcoded, e.g. it can be r
 
 Explicitly setting property values via `.ConfigureProperty<T>()` and `.InitializeHandlerProperty<T>()` has been deprecated. Instead configure the properties explicitly using:
 
-snippet: 5to6-ExplicitProperties
\ No newline at end of file
+snippet: 5to6-ExplicitProperties
+
+## IWantToRunWhenBusStartsAndStops is obsolete
+
+The `IWantToRunWhenBusStartsAndStops` interface is no longer supported. When self-hosting, call any startup code after `Endpoint.Create` or any cleanup code after `Endpoint.Stop`. When using either NServiceBus.Host or NServiceBus.Host.AzureCloudService, use the host's `IWantToRunWhenEndpointStartsAndStops` interface instead.
```
Comment (SimonCropp) [inline nservicebus/upgrades/5to6.md]: link to host doco and AzureCloudServicehost

Comment (SimonCropp) [inline nservicebus/upgrades/5to6.md]: instead is redundant

File: nservicebus/upgrades/5to6.md (hunk 3 above)
Comment (SimonCropp) [inline nservicebus/upgrades/5to6.md]: instead is redundant

File: Snippets/Snippets_6/Lifecycle/EndpointStartAndStop.cs [hunk 5]
```diff
@@ -2,19 +2,48 @@
 {
     using System.Threading.Tasks;
     using NServiceBus;
+    using NServiceBus.Features;
 
     #region lifecycle-EndpointStartAndStop
 
-    class RunWhenEndpointStartsAndStops : IWantToRunWhenEndpointStartsAndStops
+    class RunWhenEndpointStartsAndStops
```
Comment (SimonCropp) [inline Snippets/Snippets_6/Lifecycle/EndpointStartAndStop.cs]: i think move the region inside the method

File: Snippets/Snippets_6/Host/EndpointStartAndStop.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    #region host-EndpointStartAndStop
+
+    class RunWhenEndpointStartsAndStops : IWantToRunWhenEndpointStartsAndStops
+    {
+        public async Task Start(IMessageSession session)
+        {
+            // perform startup logic
+        }
+
+        public async Task Stop(IMessageSession session)
```
Comment (SimonCropp) [inline Snippets/Snippets_6/Host/EndpointStartAndStop.cs]: isnt IMessageSession the wrong type for endpoint shutdown? cc @danielmarbach @timbussmann

Comment (danielmarbach) [inline Snippets/Snippets_6/Host/EndpointStartAndStop.cs]: Why @SimonCropp ?

Comment (SimonCropp) [inline Snippets/Snippets_6/Host/EndpointStartAndStop.cs]: isnt message session for sending messages? would u want to do this on shutdown?

Comment (danielmarbach) [inline Snippets/Snippets_6/Host/EndpointStartAndStop.cs]: Inform another system about your shutdown, heartbeats, etc. That was the initial reasoning when we introduced it.

Comment (SimonCropp) [inline Snippets/Snippets_6/Host/EndpointStartAndStop.cs]: seems weird to queue more messages during the shutdown sequence

Comment (danielmarbach) [inline Snippets/Snippets_6/Host/EndpointStartAndStop.cs]: We talk about one endpoint being shut down. Not the whole system.

File: nservicebus/hosting/critical-errors.md (hunk 2 above)
Comment (colin-higgins) [inline nservicebus/hosting/critical-errors.md]: @SimonCropp We are planning on bringing snippets in for each version

File: Snippets/Snippets_4/Host/EndpointStartAndStop.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
+﻿namespace Snippets4.Host
+{
+    using NServiceBus;
+
+    #region host-EndpointStartAndStop
```
Comment (indualagarsamy) [inline Snippets/Snippets_4/Host/EndpointStartAndStop.cs]: @colin-higgins - why do we need this snippet. We already have this? snippet:lifecycle-EndpointStartAndStop

File: Snippets/Snippets_5/Host/EndpointStartAndStop.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
+﻿namespace Snippets5.Host
+{
+    using NServiceBus;
+
+    #region host-EndpointStartAndStop
```
Comment (indualagarsamy) [inline Snippets/Snippets_5/Host/EndpointStartAndStop.cs]: Same as above. Thought we have this snippet:lifecycle-EndpointStartAndStop

**Comment by indualagarsamy** [comment]
@colin-higgins @boblangley @kbaley - For the documentation items that aren't complete that targets different documentation pages, can you open separate PRs for each doco page? Small PRs are much easier to review and we can merge them sooner and hold the entire PR holding on a different change on a different page.

File: nservicebus/upgrades/5to6.md [hunk 9]
```diff
@@ -664,3 +664,9 @@ NOTE: This ID needs to be stable and should never be hardcoded, e.g. it can be r
 Explicitly setting property values via `.ConfigureProperty<T>()` and `.InitializeHandlerProperty<T>()` has been deprecated. Instead configure the properties explicitly using:
 
 snippet: 5to6-ExplicitProperties
+
+## IWantToRunWhenBusStartsAndStops is obsolete
+
+The `IWantToRunWhenBusStartsAndStops` interface is no longer supported. When self-hosting, call any startup code after `Endpoint.Create` or any cleanup code after `Endpoint.Stop`. When using either NServiceBus.Host or NServiceBus.Host.AzureCloudService, use the host's `IWantToRunWhenEndpointStartsAndStops` interface.
```
Comment (indualagarsamy) [inline nservicebus/upgrades/5to6.md]: Is there any associated documentation you can link to on the NServiceBus.Host?

Comment (indualagarsamy) [inline nservicebus/upgrades/5to6.md]: Also, this section needs to be more described here: http://docs.particular.net/nservicebus/upgrades/5to6#endpoint-start-and-stop

File: nservicebus/upgrades/5to6.md [hunk 10]
```diff
@@ -664,3 +664,9 @@ NOTE: This ID needs to be stable and should never be hardcoded, e.g. it can be r
...
+
+For tests that implement `IWantToRunWhenBusStartsAndStops`, use a `FeatureStartupTask`.
```
Comment (indualagarsamy) [inline nservicebus/upgrades/5to6.md]: Is there documentation to FeatureStartupTask or an associated sample to link to?

File: nservicebus/lifecycle/endpointstartandstop.md [hunk 11]
```diff
@@ -72,6 +72,7 @@ When the Bus is disposed, all instances of `IWantToRunWhenBusStartsAndStops` are
 
 NOTE: `Stop()` will wait for any outstanding instances of `Start()` to complete. If an instance of `IWantToRunWhenBusStartsAndStops` needs to be long running then it must start it's own background thread. Failure to do so will prevent the bus from being disposed.
 
+NOTE: `Stop()` is not supported in Version 3
```
Comment (indualagarsamy) [inline nservicebus/lifecycle/endpointstartandstop.md]: Rather than use an additional note, can this not be added to the previous note? I'd phrase it as:

`Stop()` method is not supported in Versions 3 and below. 

For Version language, please see: https://github.com/particular/docs.particular.net#version-language

**Comment by boblangley** [comment]
Split out remaining changes into separate PRs to make review easier.

File: nservicebus/hosting/critical-errors.md (hunk 2 above)
Comment (boblangley) [inline nservicebus/hosting/critical-errors.md]: I split these changes out into a separate PR #1382

File: Snippets/Snippets_6/Lifecycle/EndpointStartAndStop.cs (hunk 5 above)
Comment (boblangley) [inline Snippets/Snippets_6/Lifecycle/EndpointStartAndStop.cs]: We just removed the FeatureStartupTask part. We will just point that to the article with it's existing snippets

File: Snippets/Snippets_6/Lifecycle/EndpointStartAndStop.cs (hunk 1 above)
Comment (boblangley) [inline Snippets/Snippets_6/Lifecycle/EndpointStartAndStop.cs]: I removed this part in favor of pointing them to the article and it's existing snippets.

File: Snippets/Snippets_4/Host/EndpointStartAndStop.cs (hunk 7 above)
Comment (boblangley) [inline Snippets/Snippets_4/Host/EndpointStartAndStop.cs]: We split the snippets, one to show generic lifecycle presented by the core (lifecycle-EndpointStartStop) and lifecycle usable in the host (host-EndpointStartAndStop)

File: nservicebus/lifecycle/endpointstartandstop.md (hunk 11 above)
Comment (kbaley) [inline nservicebus/lifecycle/endpointstartandstop.md]: Reverting this change and including it in https://github.com/Particular/docs.particular.net/pull/1381

File: nservicebus/lifecycle/endpointstartandstop.md [hunk 12]
```diff
@@ -10,43 +10,45 @@ redirects:
...
+Classes that plug into the startup/shutdown sequence are invoked just after the endpoint instance has been started and just before it is stopped. Use this approach for any tasks that need to execute with the same life-cycle as the endpoint instance.
 
 NOTE: The endpoint instance keeps an internal list of instances which need to be stopped but the instances are registered with the `Builder` as Instance Per Call by default. This means that any attempt to resolve a custom instance via dependency injection will result in a new instance and not the one which has been started and which will be stopped.
 
-
 ## Versions 6 and above
 
+The `IWantToRunWhenEndpointStartsAndStops` interface was previously available within the NServiceBus core library as `IWantToRunWhenBusStartsAndStops`. This interface is only available for those using the [NServiceBus.Host](https://www.nuget.org/packages/NServiceBus.Host/). For more information about the NServiceBus.Host, see the [documentation for the host](/nservicebus/hosting/nservicebus-host).
```
Comment (indualagarsamy) [inline nservicebus/lifecycle/endpointstartandstop.md]: The `IWantToRunWhenEndpointStartsAndStops` is a new interface. It wasn't previously available in the Core. What was previously available was the `IWantToRunWhenBusStartsAndStops ` interface.

**Comment by indualagarsamy** [comment]
After reviewing this PR, here are the following changes that I made:
1. Added a new upgrade guide for Azure Cloud Service Host v6 to 7 and documented the new interface addition there. cc / @SeanFeldman 
2. Changed the v5to6 upgrade guide and removed the extra information about interface startup behavior and instead just linking to the relevant upgrade guides. cc / @kbaley 
3. Removed the new interface startup behavior description from the endpointstartandstop.md to both the NServiceBus Host and Azure Cloud Service Host documentation. While this is an information repeat, essentially both hosts can potentially choose to have different behaviors and this belongs more in the host rather than the core documentation. cc / @colin-higgins @boblangley @SeanFeldman 

@SimonCropp - Can you review this please?
